File: scripts/generate_embeddings.py
#!/usr/bin/env python3
"""
Embedding Generation Script for RAG System
Reads extracted chunks and generates embeddings using OpenAI API.
"""

import os
import json
import logging
import time
from typing import List, Dict, Any, Optional
from dataclasses import dataclass
import argparse

try:
    from openai import OpenAI
except ImportError:
    print("Error: openai library not installed. Install with: pip install openai")
    exit(1)

# Configure logging
logging.basicConfig(
    level=logging.INFO,
    format='%(asctime)s - %(levelname)s - %(message)s',
    handlers=[
        logging.FileHandler('embedding_generation.log'),
        logging.StreamHandler()
    ]
)
logger = logging.getLogger(__name__)


class EmbeddingGenerator:
    """Generates embeddings for text chunks using OpenAI API"""

    def __init__(self, api_key: str, model: str = "text-embedding-3-small", batch_size: int = 100):
        """
        Initialize the embedding generator

        Args:
            api_key: OpenAI API key
            model: Embedding model to use
            batch_size: Number of chunks to process in each batch
        """
        self.client = OpenAI(api_key=api_key)
        self.model = model
        self.batch_size = batch_size
        self.total_tokens_used = 0
        logger.info(f"Initialized EmbeddingGenerator (model={model}, batch_size={batch_size})")

    def generate_embedding(self, text: str) -> Optional[List[float]]:
        """
        Generate embedding for a single text

        Args:
            text: Text to embed

        Returns:
            Embedding vector or None on error
        """
        try:
            response = self.client.embeddings.create(
                input=text,
                model=self.model
            )

            # Track token usage
            self.total_tokens_used += response.usage.total_tokens

            return response.data[0].embedding

        except Exception as e:
            logger.error(f"Error generating embedding: {e}")
            return None

    def generate_embeddings_batch(self, texts: List[str]) -> List[Optional[List[float]]]:
        """
        Generate embeddings for a batch of texts

        Args:
            texts: List of texts to embed

        Returns:
            List of embedding vectors (None for failed embeddings)
        """
        if not texts:
            return []

        try:
            response = self.client.embeddings.create(
                input=texts,
                model=self.model
            )

            # Track token usage
            self.total_tokens_used += response.usage.total_tokens

            # Extract embeddings in order
            embeddings = [data.embedding for data in response.data]

            logger.info(f"Generated {len(embeddings)} embeddings (tokens used: {response.usage.total_tokens})")

            return embeddings

        except Exception as e:
            logger.error(f"Error generating batch embeddings: {e}")
            # Fall back to individual generation
            logger.info("Falling back to individual embedding generation...")
            return [self.generate_embedding(text) for text in texts]

    def process_chunks(self, chunks: List[Dict[str, Any]],
                      show_progress: bool = True,
                      rate_limit_delay: float = 0.1) -> List[Dict[str, Any]]:
        """
        Process all chunks and add embeddings

        Args:
            chunks: List of chunk dictionaries
            show_progress: Whether to show progress information
            rate_limit_delay: Delay between batches (seconds)

        Returns:
            List of chunks with embeddings added
        """
        logger.info(f"Processing {len(chunks)} chunks...")

        embedded_chunks = []
        failed_chunks = []

        # Process in batches
        for i in range(0, len(chunks), self.batch_size):
            batch_chunks = chunks[i:i + self.batch_size]
            batch_texts = [chunk['text'] for chunk in batch_chunks]

            if show_progress:
                progress = (i / len(chunks)) * 100
                logger.info(f"Progress: {progress:.1f}% ({i}/{len(chunks)} chunks)")

            # Generate embeddings for batch
            embeddings = self.generate_embeddings_batch(batch_texts)

            # Add embeddings to chunks
            for chunk, embedding in zip(batch_chunks, embeddings):
                if embedding is not None:
                    chunk_with_embedding = chunk.copy()
                    chunk_with_embedding['embedding'] = embedding
                    embedded_chunks.append(chunk_with_embedding)
                else:
                    failed_chunks.append(chunk['id'])
                    logger.warning(f"Failed to generate embedding for chunk: {chunk['id']}")

            # Rate limiting delay
            if i + self.batch_size < len(chunks):
                time.sleep(rate_limit_delay)

        logger.info(f"Successfully embedded {len(embedded_chunks)}/{len(chunks)} chunks")

        if failed_chunks:
            logger.warning(f"Failed chunks: {len(failed_chunks)}")
            logger.warning(f"Failed IDs: {failed_chunks[:10]}...")  # Show first 10

        return embedded_chunks


def load_chunks(input_path: str) -> Dict[str, Any]:
    """
    Load chunks from JSON file

    Args:
        input_path: Path to input JSON file

    Returns:
        Dictionary containing chunks and metadata
    """
    logger.info(f"Loading chunks from: {input_path}")

    if not os.path.exists(input_path):
        logger.error(f"Input file not found: {input_path}")
        raise FileNotFoundError(f"Input file not found: {input_path}")

    with open(input_path, 'r', encoding='utf-8') as f:
        data = json.load(f)

    logger.info(f"Loaded {len(data.get('chunks', []))} chunks")
    return data


def save_embeddings(output_path: str, chunks: List[Dict[str, Any]], metadata: Dict[str, Any]):
    """
    Save chunks with embeddings to JSON file

    Args:
        output_path: Path to output JSON file
        chunks: List of chunks with embeddings
        metadata: Metadata dictionary
    """
    logger.info(f"Saving embeddings to: {output_path}")

    os.makedirs(os.path.dirname(output_path), exist_ok=True)

    output_data = {
        "metadata": metadata,
        "chunks": chunks
    }

    with open(output_path, 'w', encoding='utf-8') as f:
        json.dump(output_data, f, indent=2, ensure_ascii=False)

    # Get file size
    file_size_mb = os.path.getsize(output_path) / (1024 * 1024)
    logger.info(f"Saved {len(chunks)} chunks with embeddings ({file_size_mb:.2f} MB)")


def calculate_cost(total_tokens: int, model: str = "text-embedding-3-small") -> float:
    """
    Calculate approximate cost based on token usage

    Args:
        total_tokens: Total tokens used
        model: Model name

    Returns:
        Estimated cost in USD
    """
    # Pricing as of 2025 (tokens per dollar)
    pricing = {
        "text-embedding-3-small": 0.00002,  # $0.02 per 1M tokens
        "text-embedding-3-large": 0.00013,  # $0.13 per 1M tokens
        "text-embedding-ada-002": 0.0001,   # $0.10 per 1M tokens
    }

    cost_per_token = pricing.get(model, 0.0001)
    return total_tokens * cost_per_token


def main():
    """Main execution function"""

    parser = argparse.ArgumentParser(description='Generate embeddings for document chunks')
    parser.add_argument('--input', type=str,
                       default='/Users/a21/Desktop/routellm-chatbot-railway/data/extracted_chunks.json',
                       help='Input JSON file with extracted chunks')
    parser.add_argument('--output', type=str,
                       default='/Users/a21/Desktop/routellm-chatbot-railway/data/embeddings.json',
                       help='Output JSON file for embeddings')
    parser.add_argument('--api-key', type=str,
                       help='OpenAI API key (can also use OPENAI_API_KEY env var)')
    parser.add_argument('--model', type=str, default='text-embedding-3-small',
                       help='OpenAI embedding model to use')
    parser.add_argument('--batch-size', type=int, default=100,
                       help='Batch size for API requests')
    parser.add_argument('--rate-limit-delay', type=float, default=0.1,
                       help='Delay between batches (seconds)')

    args = parser.parse_args()

    # Get API key
    api_key = args.api_key or os.environ.get('OPENAI_API_KEY')

    if not api_key:
        logger.error("OpenAI API key not provided. Use --api-key or set OPENAI_API_KEY environment variable")
        return

    logger.info("=" * 80)
    logger.info("Starting Embedding Generation Pipeline")
    logger.info("=" * 80)

    try:
        # Load chunks
        data = load_chunks(args.input)
        chunks = data.get('chunks', [])

        if not chunks:
            logger.error("No chunks found in input file")
            return

        # Initialize generator
        generator = EmbeddingGenerator(
            api_key=api_key,
            model=args.model,
            batch_size=args.batch_size
        )

        # Generate embeddings
        start_time = time.time()
        embedded_chunks = generator.process_chunks(
            chunks,
            show_progress=True,
            rate_limit_delay=args.rate_limit_delay
        )
        end_time = time.time()

        # Update metadata
        metadata = data.get('metadata', {})
        metadata.update({
            'total_chunks_with_embeddings': len(embedded_chunks),
            'embedding_model': args.model,
            'total_tokens_used': generator.total_tokens_used,
            'processing_time_seconds': round(end_time - start_time, 2),
            'embedding_dimension': len(embedded_chunks[0]['embedding']) if embedded_chunks else 0
        })

        # Calculate cost
        estimated_cost = calculate_cost(generator.total_tokens_used, args.model)
        metadata['estimated_cost_usd'] = round(estimated_cost, 4)

        # Save embeddings
        save_embeddings(args.output, embedded_chunks, metadata)

        # Print summary
        logger.info("=" * 80)
        logger.info("Embedding Generation Complete!")
        logger.info("=" * 80)
        logger.info(f"Total chunks processed: {len(embedded_chunks)}/{len(chunks)}")
        logger.info(f"Embedding model: {args.model}")
        logger.info(f"Embedding dimension: {metadata['embedding_dimension']}")
        logger.info(f"Total tokens used: {generator.total_tokens_used:,}")
        logger.info(f"Estimated cost: ${estimated_cost:.4f}")
        logger.info(f"Processing time: {metadata['processing_time_seconds']:.2f} seconds")
        logger.info(f"Output saved to: {args.output}")
        logger.info("=" * 80)

        # Print sample
        if embedded_chunks:
            logger.info("\nSample embedded chunk:")
            sample = embedded_chunks[0]
            logger.info(f"  ID: {sample['id']}")
            logger.info(f"  Text preview: {sample['text'][:100]}...")
            logger.info(f"  Embedding dimension: {len(sample['embedding'])}")
            logger.info(f"  Document type: {sample['metadata']['doc_type']}")
            logger.info(f"  Source: {sample['metadata']['filename']}")

    except Exception as e:
        logger.error(f"Error in embedding generation pipeline: {e}", exc_info=True)
        raise


if __name__ == "__main__":
    main()

File: scripts/generate-embeddings.js
#!/usr/bin/env node

/**
 * Generate Embeddings for RAG System
 *
 * Processes the existing 123 KB documents and generates vector embeddings
 * using OpenAI's text-embedding-3-small model ($0.02/1M tokens)
 *
 * This script:
 * 1. Loads all 123 documents from insurance-argumentation-kb.ts
 * 2. Chunks each document into ~500 token segments
 * 3. Generates embeddings for each chunk using OpenAI
 * 4. Stores embeddings in PostgreSQL with pgvector
 * 5. Creates HNSW indexes for fast similarity search
 */

const fs = require('fs');
const path = require('path');
const { Client } = require('pg');

// ============================================================================
// CONFIGURATION
// ============================================================================

const OPENAI_API_KEY = process.env.OPENAI_API_KEY;
const DATABASE_URL = process.env.DATABASE_URL;

// Embedding configuration
const EMBEDDING_MODEL = 'text-embedding-3-small'; // $0.02/1M tokens
const EMBEDDING_DIMENSIONS = 1536; // Default for text-embedding-3-small
const CHUNK_SIZE = 500; // Tokens per chunk (approximate)
const CHUNK_OVERLAP = 50; // Overlap between chunks

// Processing configuration
const BATCH_SIZE = 100; // Process 100 chunks at a time
const RATE_LIMIT_DELAY = 1000; // 1 second between batches

// ============================================================================
// UTILITY FUNCTIONS
// ============================================================================

/**
 * Simple text chunking by character count (approximation of tokens)
 * 1 token ≈ 4 characters for English text
 */
function chunkText(text, chunkSize = 500, overlap = 50) {
  const chunks = [];
  const chunkChars = chunkSize * 4; // Approximate characters
  const overlapChars = overlap * 4;

  let start = 0;
  let chunkIndex = 0;

  while (start < text.length) {
    const end = Math.min(start + chunkChars, text.length);
    const chunk = text.substring(start, end);

    chunks.push({
      text: chunk,
      tokens: Math.ceil(chunk.length / 4),
      chunk_index: chunkIndex,
      position_start: start,
      position_end: end,
    });

    start = end - overlapChars; // Move forward with overlap
    chunkIndex++;
  }

  // Set total_chunks for all chunks
  chunks.forEach(chunk => {
    chunk.total_chunks = chunks.length;
  });

  return chunks;
}

/**
 * Generate embeddings using OpenAI API
 */
async function generateEmbedding(text) {
  const response = await fetch('https://api.openai.com/v1/embeddings', {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      'Authorization': `Bearer ${OPENAI_API_KEY}`,
    },
    body: JSON.stringify({
      model: EMBEDDING_MODEL,
      input: text,
      dimensions: EMBEDDING_DIMENSIONS,
    }),
  });

  if (!response.ok) {
    const error = await response.text();
    throw new Error(`OpenAI API error: ${response.status} - ${error}`);
  }

  const data = await response.json();
  return data.data[0].embedding;
}

/**
 * Generate embeddings in batches with rate limiting
 */
async function generateEmbeddingsBatch(texts) {
  const embeddings = [];

  for (let i = 0; i < texts.length; i += BATCH_SIZE) {
    const batch = texts.slice(i, i + BATCH_SIZE);

    console.log(`  Processing batch ${Math.floor(i / BATCH_SIZE) + 1}/${Math.ceil(texts.length / BATCH_SIZE)} (${batch.length} chunks)`);

    const batchEmbeddings = await Promise.all(
      batch.map(text => generateEmbedding(text))
    );

    embeddings.push(...batchEmbeddings);

    // Rate limiting
    if (i + BATCH_SIZE < texts.length) {
      await new Promise(resolve => setTimeout(resolve, RATE_LIMIT_DELAY));
    }
  }

  return embeddings;
}

/**
 * Load documents from TypeScript knowledge base
 */
function loadDocuments() {
  console.log('📚 Loading documents from knowledge base...\n');

  // Import the KB module
  const kbPath = path.join(__dirname, '..', 'lib', 'insurance-argumentation-kb.ts');

  // Since it's TypeScript, we need to parse it
  // For now, we'll use a simple approach - load the compiled JS version
  // In production, we'd use ts-node or compile first

  try {
    // Try to load from compiled JS
    const { INSURANCE_KB_DOCUMENTS } = require('../lib/insurance-argumentation-kb');
    return INSURANCE_KB_DOCUMENTS;
  } catch (error) {
    console.error('❌ Could not load knowledge base');
    console.error('   Make sure the project is built: npm run build');
    console.error('   Error:', error.message);
    process.exit(1);
  }
}

// ============================================================================
// MAIN PROCESSING FUNCTION
// ============================================================================

async function generateEmbeddings() {
  console.log('🚀 RAG Embedding Generation');
  console.log('============================\n');

  // Validate environment
  if (!OPENAI_API_KEY) {
    console.error('❌ OPENAI_API_KEY environment variable not set');
    process.exit(1);
  }

  if (!DATABASE_URL) {
    console.error('❌ DATABASE_URL environment variable not set');
    console.error('   This script must run in Railway environment');
    console.error('   Use: railway run npm run rag:build');
    process.exit(1);
  }

  console.log('✓ OPENAI_API_KEY configured');
  console.log('✓ DATABASE_URL configured');
  console.log(`✓ Embedding model: ${EMBEDDING_MODEL}`);
  console.log(`✓ Chunk size: ${CHUNK_SIZE} tokens (~${CHUNK_SIZE * 4} chars)`);
  console.log(`✓ Chunk overlap: ${CHUNK_OVERLAP} tokens\n`);

  // Load documents
  const documents = loadDocuments();
  console.log(`✓ Loaded ${documents.length} documents\n`);

  // Connect to database
  console.log('🔌 Connecting to PostgreSQL...');
  const client = new Client({ connectionString: DATABASE_URL });
  await client.connect();
  console.log('✓ Connected\n');

  try {
    // Process each document
    let totalChunks = 0;
    let totalTokens = 0;

    for (let i = 0; i < documents.length; i++) {
      const doc = documents[i];
      const docNum = i + 1;

      console.log(`[${docNum}/${documents.length}] ${doc.filename}`);
      console.log(`  Category: ${doc.category}`);
      console.log(`  Title: ${doc.title}`);

      // Combine document content
      const fullContent = `${doc.title}\n\n${doc.summary}\n\n${doc.content}`;

      // Calculate hash for version tracking
      const crypto = require('crypto');
      const hash = crypto.createHash('sha256').update(fullContent).digest('hex');

      // Check if document already exists
      const existingDoc = await client.query(
        'SELECT id, hash FROM rag_documents WHERE id = $1',
        [doc.id]
      );

      if (existingDoc.rows.length > 0 && existingDoc.rows[0].hash === hash) {
        console.log('  ⏭️  Already processed (same hash) - skipping\n');
        continue;
      }

      // Insert/update document
      await client.query(`
        INSERT INTO rag_documents (id, filename, filepath, type, content, summary, metadata, hash)
        VALUES ($1, $2, $3, $4, $5, $6, $7, $8)
        ON CONFLICT (id) DO UPDATE SET
          content = EXCLUDED.content,
          summary = EXCLUDED.summary,
          metadata = EXCLUDED.metadata,
          hash = EXCLUDED.hash,
          version = rag_documents.version + 1,
          updated_at = NOW()
      `, [
        doc.id,
        doc.filename,
        doc.filename, // filepath same as filename for KB docs
        'docx', // Default type
        fullContent,
        doc.summary,
        JSON.stringify({
          category: doc.category,
          scenarios: doc.scenarios || [],
          states: doc.applicableStates || [],
          tags: doc.tags || [],
        }),
        hash,
      ]);

      // Chunk the document
      const chunks = chunkText(fullContent, CHUNK_SIZE, CHUNK_OVERLAP);
      console.log(`  Chunks: ${chunks.length}`);

      totalChunks += chunks.length;
      totalTokens += chunks.reduce((sum, c) => sum + c.tokens, 0);

      // Generate embeddings
      console.log(`  Generating embeddings...`);
      const chunkTexts = chunks.map(c => c.text);
      const embeddings = await generateEmbeddingsBatch(chunkTexts);

      // Insert chunks with embeddings
      console.log(`  Storing in database...`);

      for (let j = 0; j < chunks.length; j++) {
        const chunk = chunks[j];
        const embedding = embeddings[j];
        const chunkId = `${doc.id}_chunk_${j}`;

        await client.query(`
          INSERT INTO rag_chunks (
            id, document_id, text, tokens, chunk_index, total_chunks,
            position_start, position_end, embedding
          )
          VALUES ($1, $2, $3, $4, $5, $6, $7, $8, $9)
          ON CONFLICT (id) DO UPDATE SET
            text = EXCLUDED.text,
            tokens = EXCLUDED.tokens,
            embedding = EXCLUDED.embedding,
            updated_at = NOW()
        `, [
          chunkId,
          doc.id,
          chunk.text,
          chunk.tokens,
          chunk.chunk_index,
          chunk.total_chunks,
          chunk.position_start,
          chunk.position_end,
          `[${embedding.join(',')}]`, // pgvector format
        ]);
      }

      console.log('  ✅ Complete\n');

      // Progress update every 10 documents
      if (docNum % 10 === 0) {
        console.log(`📊 Progress: ${((docNum / documents.length) * 100).toFixed(1)}% (${docNum}/${documents.length})\n`);
      }
    }

    // Create HNSW index for fast vector search
    console.log('🔍 Creating HNSW index for vector search...');
    await client.query(`
      CREATE INDEX IF NOT EXISTS idx_chunks_embedding_hnsw
      ON rag_chunks USING hnsw (embedding vector_cosine_ops)
      WITH (m = 16, ef_construction = 64)
    `);
    console.log('✓ HNSW index created\n');

    // Final statistics
    const docCount = await client.query('SELECT COUNT(*) FROM rag_documents');
    const chunkCount = await client.query('SELECT COUNT(*) FROM rag_chunks');

    console.log('═══════════════════════════════════════');
    console.log('✨ EMBEDDING GENERATION COMPLETE!');
    console.log('═══════════════════════════════════════\n');

    console.log('📊 Statistics:');
    console.log(`   Documents: ${docCount.rows[0].count}`);
    console.log(`   Chunks: ${chunkCount.rows[0].count}`);
    console.log(`   Total tokens: ~${totalTokens.toLocaleString()}`);
    console.log(`   Estimated cost: $${((totalTokens / 1_000_000) * 0.02).toFixed(4)}`);

    console.log('\n🔜 Next steps:');
    console.log('   1. ✅ Embeddings generated and stored');
    console.log('   2. ✅ HNSW index created for fast search');
    console.log('   3. 🧪 Test RAG retrieval: npm run test:rag');
    console.log('   4. 🔗 Integrate with chat API\n');

  } catch (error) {
    console.error('\n❌ Error generating embeddings:');
    console.error(error);
    process.exit(1);
  } finally {
    await client.end();
  }
}

// Run the script
generateEmbeddings();
